packages/dds/tree/src/util/rangeMap.ts

```diff
@@ -227,3 +228,108 @@ export function deleteFromRangeMap<T>(map: RangeMap<T>, start: number, length: n
 		}
 	}
 }
+
+/**
+ * Given a query range, partition the range map into multiple segments according to the following rules:
+ *
+ * 1. If a range entry overlaps with the query range, store the overlapping portion as a segment in the result.
+ * 2. If there is an "empty" space between two consecutive range entries, store the range between the end
+ * of the first entry and the start of the second entry as a new segment with an undefined value.
+ *
+ * For example, consider a range map:
+ *
+ * map = [[start: 1, length: 3, value: 'a'], [start: 6, length: 2, value: 'b']]
+ *
+ * When partitioning the range map within the query range [0, 7], the result will be:
+ *
+ * [[ start: 0, length: 1, value: undefined ], [ start: 1, length: 3, value: 'a' ],
+ * [ start: 4, length: 2, value: undefined ], [ start: 6, length: 1, value: 'b' ]]
+ */
+export function partitionAllRangesWithinMap<T>(
+	map: RangeMap<T>,
+	start: number,
+	length: number,
+): RangeFullQueryResult<T>[] {
```

Comment:
@yann-achard-MS I got what you mean, when we group empty slots between two ranges as a new `rangeEntry` with a value of "undefined," it's reasonable to consider the risk in `rangeMap` when T itself can be undefined. However, within the current `detachedFieldIndex`, I guess the value of a node range entry will only be a `ForestRootId` or undefined, and within the function `buildTree`, we will always create a new root/field for a contiguous range of nodes if they do not have an existing root assigned to them (i.e. their existing root is undefined)

So I think we could move the logic to create the `rangeEntry` with an undefined value from `rangeMap` to `detachedFieldIndex`, if we assume it is risky in general usage of `rangeMap`, but reasonable in `detachedFieldIndex`. May I know your thoughts?

Reply:
If I understand correctly, you're making the case that the drawbacks of the current API mostly arise out of `T` possibly being undefined, and since the `DetachedFieldIndex` class knows that `T = ForestRootId`, it could still offer that API. In practice, this would mean that `DetachedFieldIndex.partitionDetachedNodeRanges` would fill in the gaps (if any) with instances of `RangeQueryResult<undefined>`. Am I understanding you right?

I think your point is valid. It doesn't address all the drawbacks of the current API, but we could still embrace it and address the drawbacks some other way. I think that would likely involve making `DetachedFieldIndex.partitionDetachedNodeRanges` return something like `{ count: number, value?: ForestRootId }[]`.
How does that sound to you?

Reply:
Yeah this sounds good, btw we also need to include the start (the start node id of a contiguous range) as the result, is that correct?

Reply:
Not for `DetachedFieldIndex.partitionDetachedNodeRanges` because each range starts where the previous one ends, and the first range starts where the queried range starts.
So, to recap, I would expect:
`getAllEntriesFromMap` to return `{ value: T; start: number; length: number }[]` <- notice there is no `T | undefined ` here
`DetachedFieldIndex.partitionDetachedNodeRanges` to return `{ count: number, value?: ForestRootId }[]`.

Reply:
Updated in https://github.com/microsoft/FluidFramework/pull/20515/commits/65a13fc7b18627056797cbff36305c0b23f35529